File: services/project/internal/store/crypto.go
package store

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/axebom/axebom/libs/go-shared/platform/db"
)

// ---------------------------------------------------------------------------
// Crypto assets — normalize.crypto_assets, read cross-schema.
//
// ⚠ SAME PATTERN AS ListDependencies (dependencies.go): this service owns
// project.* only, not normalize.*. One query, no cross-schema SQL JOIN
// (CLAUDE.md invariant 11), scoped by the project's current CBOM
// bom_document via resolveCurrentBOMDocument. The engines list reads
// normalize.crypto_asset_provenance in the same statement — same schema, so
// invariant 11 does not apply, and RLS scopes both tables.
//
// ⚠ THIS IS THE INTERACTIVE READ PATH, DISTINCT FROM services/report'S
// RENDERING. The report service (render.CryptoAsset / CBOMSheets) builds a
// downloadable XLSX/PDF/JSON artifact for one point in time. This is the
// live "Crypto inventory" screen the CBOM sidebar section links to — same
// underlying table, same type-discrimination discipline, a different
// consumer with a different shape need (JSON over HTTP, not a workbook).
// ---------------------------------------------------------------------------

// CryptoAsset is one row of the crypto inventory screen
// (frontend/src/routes/boms/CryptoInventory.tsx's CryptoAssetRow).
//
// ⚠ TYPE-DISCRIMINATED, NOT ONE FLAT COLUMN SET RENDERED FOR EVERY TYPE.
// CERT-In Table 9 gives Algorithms/Keys/Protocols/Certificates different
// field sets (CLAUDE.md invariant 5); the JSON omits a field entirely
// (`omitempty`) rather than send an empty string for a column that asset
// type never had, so the frontend can tell "not applicable to this type"
// from "applicable and not-provided" the same way the Go struct already
// does by leaving the Go field its zero value only where the DB column is
// NULL.
//
// ⚠ THE IDENTITY AND EVIDENCE FIELDS ARE NEVER OMITTED. They apply to every
// asset type, so an empty `[]` / `{}` is an answer ("no engine reported a
// location"), and a consumer can rely on the key being there.
type CryptoAsset struct {
	// AssetKey is the asset's identity — the docs/03-NORMALIZER-SPEC.md §1.6
	// merge key (`algorithm:…`, `key:fp:sha256:…`, `cert:…`, `protocol:…`,
	// `opaque:…`). Unlike the row id it survives re-normalization, so it is
	// what a QBOM references and what the CBOM export uses as the bom-ref.
	AssetKey string `json:"asset_key"`
	// IdentityRule is the §1.6 ladder rule that produced AssetKey, and
	// IdentityConfidence how much weight it bears (high / medium / low).
	IdentityRule       string `json:"identity_rule"`
	IdentityConfidence string `json:"identity_confidence"`

	ComponentKey string `json:"component_key,omitempty"`
	AssetType    string `json:"asset_type"`
	Name         string `json:"name"`

	// ---- algorithm ----
	Primitive              string   `json:"primitive,omitempty"`
	Mode                   string   `json:"mode,omitempty"`
	CryptoFunctions        []string `json:"crypto_functions,omitempty"`
	ClassicalSecurityLevel *int     `json:"classical_security_level,omitempty"`
	AlgorithmList          []string `json:"algorithm_list,omitempty"`

	// ---- key ----
	KeyID          string `json:"key_id,omitempty"`
	KeyState       string `json:"key_state,omitempty"`
	KeySize        *int   `json:"key_size,omitempty"`
	CreationDate   string `json:"creation_date,omitempty"`
	ActivationDate string `json:"activation_date,omitempty"`

	// ---- protocol ----
	ProtocolVersion string   `json:"protocol_version,omitempty"`
	CipherSuites    []string `json:"cipher_suites,omitempty"`

	// ---- shared by algorithm and protocol ----
	OID string `json:"oid,omitempty"`

	// ---- certificate ----
	CertSubject         string `json:"cert_subject,omitempty"`
	CertIssuer          string `json:"cert_issuer,omitempty"`
	NotValidBefore      string `json:"not_valid_before,omitempty"`
	NotValidAfter       string `json:"not_valid_after,omitempty"`
	SignatureAlgoRef    string `json:"signature_algo_ref,omitempty"`
	SubjectPublicKeyRef string `json:"subject_public_key_ref,omitempty"`
	CertFormat          string `json:"cert_format,omitempty"`
	CertExtension       string `json:"cert_extension,omitempty"`

	// ---- AxeBOM analysis — NOT CERT-In fields, excluded from coverage
	// (migrations/normalize/0003, 0005) ----
	QuantumVulnerable     bool   `json:"quantum_vulnerable"`
	QuantumFamily         string `json:"quantum_family,omitempty"`
	QuantumReadinessGroup string `json:"quantum_readiness_group,omitempty"`
	QuantumRationale      string `json:"quantum_rationale,omitempty"`
	GroverNote            string `json:"grover_note,omitempty"`
	EffectiveQuantumBits  *int   `json:"effective_quantum_bits,omitempty"`
	PQCRecommendation     string `json:"pqc_recommendation,omitempty"`
	DeprecationStatus     string `json:"deprecation_status,omitempty"`
	DeprecationRationale  string `json:"deprecation_rationale,omitempty"`
	DeprecationReference  string `json:"deprecation_reference,omitempty"`

	// ---- evidence and provenance — never scored (migrations/normalize/0019,
	// 0020) ----

	// Evidence is where engines saw the asset, repository-relative and
	// verbatim. Location is evidence, never identity: one certificate in two
	// folders is one asset with two entries here.
	Evidence []CryptoEvidence `json:"evidence"`
	// Attributes are the non-CERT-In facts the normalizer kept — padding,
	// curve, parameter set, NIST quantum category, key-material type, whether
	// a private key was committed. Passed through as stored.
	Attributes map[string]any `json:"attributes"`
	// Derivations names the CERT-In columns AxeBOM filled from a cited
	// reference table rather than an engine reporting them:
	// {column: reference_id}.
	Derivations map[string]string `json:"derivations"`
	// Engines are the distinct engines that reported this asset, sorted — read
	// from normalize.crypto_asset_provenance, one row per (asset, engine).
	Engines []string `json:"engines"`
}

// CryptoEvidence is one place an engine saw a crypto asset.
//
// Line is null when the engine reported a file but no line — a certificate
// file, a key file. Never 0 for "unknown": 0 is not a line.
type CryptoEvidence struct {
	Path   string `json:"path"`
	Line   *int   `json:"line"`
	Engine string `json:"engine"`
}

// ListCryptoAssets returns the project's current CBOM crypto-asset inventory.
func (s *Store) ListCryptoAssets(ctx context.Context, tenantID, projectID string) ([]CryptoAsset, error) {
	out := []CryptoAsset{}
	err := s.pool.WithTenant(ctx, tenantID, func(ctx context.Context, tx db.Tx) error {
		if err := projectExists(ctx, tx, projectID); err != nil {
			return err
		}

		docID, err := resolveCurrentBOMDocument(ctx, tx, projectID, "CBOM")
		if err != nil {
			return err
		}
		if docID == "" {
			return nil // no CBOM normalized yet — an honest empty list
		}

		// ⚠ THE ENGINES ARE A CORRELATED SUBQUERY, NOT A SECOND ROUND TRIP PER
		// ROW. One statement returns every asset with its engine list, so the
		// screen costs one query however large the inventory is.
		rows, err := tx.Query(ctx, `
			SELECT a.asset_key, a.identity_rule, a.identity_confidence,
			       a.component_key, a.asset_type, a.name,
			       a.primitive, a.mode, a.crypto_functions, a.classical_security_level, a.algorithm_list,
			       a.key_id, a.key_state, a.key_size,
			       to_char(a.creation_date, 'YYYY-MM-DD'), to_char(a.activation_date, 'YYYY-MM-DD'),
			       a.protocol_version, a.cipher_suites, a.oid,
			       a.cert_subject, a.cert_issuer,
			       to_char(a.not_valid_before, 'YYYY-MM-DD"T"HH24:MI:SS"Z"'),
			       to_char(a.not_valid_after, 'YYYY-MM-DD"T"HH24:MI:SS"Z"'),
			       a.signature_algo_ref, a.subject_public_key_ref, a.cert_format, a.cert_extension,
			       a.quantum_vulnerable, a.quantum_family, a.quantum_readiness_group,
			       a.quantum_rationale, a.grover_note, a.effective_quantum_bits,
			       a.pqc_recommendation, a.deprecation_status, a.deprecation_rationale,
			       a.deprecation_reference,
			       a.evidence, a.attributes, a.derivations,
			       COALESCE((SELECT array_agg(DISTINCT p.engine_id ORDER BY p.engine_id)
			                   FROM normalize.crypto_asset_provenance p
			                  WHERE p.crypto_asset_id = a.id), '{}')
			  FROM normalize.crypto_assets a
			 WHERE a.bom_document_id = $1
			 -- ⚠ id LAST, SO THE ORDER IS TOTAL. Two assets can share a type and
			 -- a name (theia reports two RSA algorithms from one certificate), and
			 -- without a tiebreak their order changed between requests.
			 ORDER BY a.asset_type, a.name, a.id`, docID)
		if err != nil {
			return fmt.Errorf("list crypto assets: %w", err)
		}
		defer rows.Close()

		for rows.Next() {
			var (
				a                                            CryptoAsset
				assetKey, identityRule, identityConfidence   *string
				componentKey                                 *string
				primitive, mode                              *string
				cryptoFunctions, algorithmList               []string
				keyID, keyState                              *string
				creationDate, activationDate                 *string
				protocolVersion                              *string
				cipherSuites                                 []string
				oid                                          *string
				certSubject, certIssuer                      *string
				notValidBefore, notValidAfter                *string
				sigRef, pubKeyRef                            *string
				certFormat, certExtension                    *string
				quantumFamily, quantumGroup                  *string
				quantumRationale, groverNote                 *string
				pqcRecommendation                            *string
				deprecationStatus                            *string
				deprecationRationale                         *string
				deprecationReference                         *string
				evidenceJSON, attributesJSON, derivationJSON []byte
				engines                                      []string
			)
			if err := rows.Scan(
				&assetKey, &identityRule, &identityConfidence,
				&componentKey, &a.AssetType, &a.Name,
				&primitive, &mode, &cryptoFunctions, &a.ClassicalSecurityLevel, &algorithmList,
				&keyID, &keyState, &a.KeySize,
				&creationDate, &activationDate,
				&protocolVersion, &cipherSuites, &oid,
				&certSubject, &certIssuer, &notValidBefore, &notValidAfter,
				&sigRef, &pubKeyRef, &certFormat, &certExtension,
				&a.QuantumVulnerable, &quantumFamily, &quantumGroup,
				&quantumRationale, &groverNote, &a.EffectiveQuantumBits,
				&pqcRecommendation, &deprecationStatus, &deprecationRationale,
				&deprecationReference,
				&evidenceJSON, &attributesJSON, &derivationJSON,
				&engines,
			); err != nil {
				return fmt.Errorf("scan crypto asset: %w", err)
			}

			// ⚠ A MALFORMED COLUMN FAILS THE REQUEST RATHER THAN BEING DROPPED.
			// All three are jsonb written only by the normalizer, so a decode
			// failure is corruption — and a derivation label silently lost would
			// present AxeBOM's lookup as an engine's claim.
			if err := decodeCryptoJSON(evidenceJSON, &a.Evidence); err != nil {
				return fmt.Errorf("decode crypto asset evidence: %w", err)
			}
			if err := decodeCryptoJSON(attributesJSON, &a.Attributes); err != nil {
				return fmt.Errorf("decode crypto asset attributes: %w", err)
			}
			if err := decodeCryptoJSON(derivationJSON, &a.Derivations); err != nil {
				return fmt.Errorf("decode crypto asset derivations: %w", err)
			}
			if a.Evidence == nil {
				a.Evidence = []CryptoEvidence{}
			}
			if a.Attributes == nil {
				a.Attributes = map[string]any{}
			}
			if a.Derivations == nil {
				a.Derivations = map[string]string{}
			}
			a.Engines = engines
			if a.Engines == nil {
				a.Engines = []string{}
			}

			a.AssetKey = deref(assetKey)
			a.IdentityRule = deref(identityRule)
			a.IdentityConfidence = deref(identityConfidence)
			a.ComponentKey = deref(componentKey)
			a.Primitive = deref(primitive)
			a.Mode = deref(mode)
			a.CryptoFunctions = cryptoFunctions
			a.AlgorithmList = algorithmList
			a.KeyID = deref(keyID)
			a.KeyState = deref(keyState)
			a.CreationDate = deref(creationDate)
			a.ActivationDate = deref(activationDate)
			a.ProtocolVersion = deref(protocolVersion)
			a.CipherSuites = cipherSuites
			a.OID = deref(oid)
			a.CertSubject = deref(certSubject)
			a.CertIssuer = deref(certIssuer)
			a.NotValidBefore = deref(notValidBefore)
			a.NotValidAfter = deref(notValidAfter)
			a.SignatureAlgoRef = deref(sigRef)
			a.SubjectPublicKeyRef = deref(pubKeyRef)
			a.CertFormat = deref(certFormat)
			a.CertExtension = deref(certExtension)
			a.QuantumFamily = deref(quantumFamily)
			a.QuantumReadinessGroup = deref(quantumGroup)
			a.QuantumRationale = deref(quantumRationale)
			a.GroverNote = deref(groverNote)
			a.PQCRecommendation = deref(pqcRecommendation)
			a.DeprecationStatus = deref(deprecationStatus)
			a.DeprecationRationale = deref(deprecationRationale)
			a.DeprecationReference = deref(deprecationReference)

			out = append(out, a)
		}
		return rows.Err()
	})
	return out, err
}

// decodeCryptoJSON unmarshals one jsonb column; NULL leaves v untouched.
func decodeCryptoJSON(raw []byte, v any) error {
	if len(raw) == 0 {
		return nil
	}
	return json.Unmarshal(raw, v)
}

File: frontend/src/lib/crypto.ts
/**
 * CBOM crypto-asset inventory — GET /v1/projects/{id}/crypto-assets.
 *
 * ⚠ TYPE-DISCRIMINATED, MIRRORING services/project/internal/store/crypto.go.
 * CERT-In Table 9 gives Algorithms/Keys/Protocols/Certificates different
 * field sets (CLAUDE.md invariant 5) — a field absent from a row means "not
 * applicable to this asset type", not "not provided". `CryptoInventory.tsx`
 * must render each type's own columns, never one flat table with every
 * field for every row.
 *
 * ⚠ EVERY IDENTITY, EVIDENCE AND PROVENANCE FIELD IS OPTIONAL. An older backend
 * returns none of them, and the page must still render — a missing location is
 * shown as missing, never invented.
 */

import { useQuery } from '@tanstack/react-query';
import { request } from './api';

/** Where one engine saw an asset: repository-relative path, line when known. */
export interface CryptoEvidence {
  path: string;
  line: number | null;
  engine: string;
}

export interface CryptoAsset {
  // ---- identity (docs/03-NORMALIZER-SPEC.md §1.6) ----
  /** The stable merge key — the same asset across scans and re-normalizations. */
  asset_key?: string;
  identity_rule?: string;
  identity_confidence?: 'high' | 'medium' | 'low';
  /** A link to an SBOM component. Empty for every current engine. */
  component_key?: string;

  asset_type: 'algorithm' | 'key' | 'protocol' | 'certificate';
  name: string;

  // ---- algorithm ----
  primitive?: string;
  mode?: string;
  crypto_functions?: string[];
  classical_security_level?: number;
  algorithm_list?: string[];

  // ---- key ----
  key_id?: string;
  key_state?: string;
  key_size?: number;
  creation_date?: string;
  activation_date?: string;

  // ---- protocol ----
  protocol_version?: string;
  cipher_suites?: string[];

  // ---- shared by algorithm and protocol ----
  oid?: string;

  // ---- certificate ----
  cert_subject?: string;
  cert_issuer?: string;
  not_valid_before?: string;
  not_valid_after?: string;
  signature_algo_ref?: string;
  subject_public_key_ref?: string;
  cert_format?: string;
  cert_extension?: string;

  // ---- evidence and provenance — never scored ----
  /** Every location every engine saw this asset at. */
  evidence?: CryptoEvidence[];
  /** The distinct engines that found it (cbomkit-theia, cbomkit-action, cdxgen-cbom). */
  engines?: string[];
  /**
   * Non-CERT-In facts: padding, curve, material type, surfaces, NIST quantum
   * category, and `private_key_in_source` / `secret_key_in_source`.
   */
  attributes?: Record<string, unknown>;
  /**
   * `{column: reference_id}` — CERT-In values AxeBOM filled from a cited
   * reference table rather than an engine reporting them. They count as
   * present and are always labelled as derived.
   */
  derivations?: Record<string, string>;

  // ---- AxeBOM analysis — not CERT-In fields, excluded from coverage ----
  quantum_vulnerable: boolean;
  quantum_family?: string;
  quantum_readiness_group?: 'vulnerable' | 'grover_note' | 'post_quantum' | 'unassessed';
  quantum_rationale?: string;
  grover_note?: string;
  effective_quantum_bits?: number;
  pqc_recommendation?: string;
  // `unassessed`: no verdict could be reached — never to be shown as current.
  deprecation_status?: 'current' | 'deprecated' | 'weak' | 'broken' | 'unassessed';
  deprecation_rationale?: string;
  deprecation_reference?: string;
}

export function useCryptoAssets(projectId: string | undefined) {
  return useQuery({
    queryKey: ['crypto-assets', projectId],
    queryFn: ({ signal }) =>
      request<{ crypto_assets: CryptoAsset[] }>(`/v1/projects/${projectId}/crypto-assets`, {
        signal,
      }),
    enabled: Boolean(projectId),
    staleTime: 30_000,
  });
}

/** The four CERT-In Table 9 asset types, in the order the inventory groups them. */
export const CRYPTO_ASSET_TYPES = ['algorithm', 'key', 'protocol', 'certificate'] as const;

/** readinessLabel names a quantum_readiness_group bucket for display. */
export function readinessLabel(group: CryptoAsset['quantum_readiness_group']): string {
  switch (group) {
    case 'vulnerable':
      return 'Quantum-vulnerable';
    case 'grover_note':
      return 'Grover note (sizing, not a break)';
    case 'post_quantum':
      return 'Already post-quantum';
    case 'unassessed':
      return 'Unassessed';
    default:
      return 'Not classified';
  }
}

/**
 * deprecationLabel names a deprecation verdict for display.
 *
 * ⚠ `unassessed` IS ITS OWN WORD, NEVER "Current". It means no verdict could be
 * reached — an RSA key whose size nobody reported is the likeliest legacy
 * 1024-bit key — and a reader filtering for problems must still see it.
 */
export function deprecationLabel(status: CryptoAsset['deprecation_status']): string {
  switch (status) {
    case 'current':
      return 'Current';
    case 'deprecated':
      return 'Deprecated';
    case 'weak':
      return 'Weak';
    case 'broken':
      return 'Broken';
    case 'unassessed':
      return 'Unassessed';
    default:
      return 'Not assessed';
  }
}

/** deprecationTone maps a verdict onto the pill palette. Only `current` is ok. */
export function deprecationTone(
  status: CryptoAsset['deprecation_status'],
): 'ok' | 'warn' | 'down' | 'info' {
  switch (status) {
    case 'current':
      return 'ok';
    case 'deprecated':
    case 'unassessed':
      return 'warn';
    case 'weak':
    case 'broken':
      return 'down';
    default:
      return 'info';
  }
}

/** formatEvidence renders one location as `path:line`, or `path` with no line. */
export function formatEvidence(evidence: CryptoEvidence): string {
  return evidence.line ? `${evidence.path}:${evidence.line}` : evidence.path;
}

export interface LocationSummary {
  /** The first distinct location. */
  primary: string;
  /** Every other distinct location, in order. */
  rest: string[];
  /** What the cell shows: the primary, plus "+N more" when there are others. */
  label: string;
}

/**
 * summarizeLocations reduces an asset's evidence to what one table cell can
 * show, keeping the rest for a tooltip. Two engines reporting the same
 * `path:line` are one location. Returns null when there is no evidence at all
 * — an older backend, or an engine that reported none — so the caller shows
 * the gap rather than a made-up place.
 */
export function summarizeLocations(
  evidence: readonly CryptoEvidence[] | undefined,
): LocationSummary | null {
  const seen: string[] = [];
  for (const item of evidence ?? []) {
    if (!item?.path) continue;
    const formatted = formatEvidence(item);
    if (!seen.includes(formatted)) seen.push(formatted);
  }
  if (seen.length === 0) return null;
  const [primary, ...rest] = seen as [string, ...string[]];
  return {
    primary,
    rest,
    label: rest.length > 0 ? `${primary} +${rest.length} more` : primary,
  };
}

/** derivedFrom says which cited reference a column's value came from, if any. */
export function derivedFrom(asset: CryptoAsset, column: string): string | undefined {
  const reference = asset.derivations?.[column];
  return typeof reference === 'string' && reference !== '' ? reference : undefined;
}

/**
 * assetEngines lists the engines that found an asset: the backend's list when
 * it sends one, otherwise the engines named in its evidence. Sorted, distinct.
 */
export function assetEngines(asset: CryptoAsset): string[] {
  const named = asset.engines?.length
    ? asset.engines
    : (asset.evidence ?? []).map((e) => e?.engine).filter(Boolean);
  return [...new Set(named)].sort();
}

/**
 * privateKeyInSource is true only when the normalizer said so — an engine that
 * reads key FILES found a private key in the scanned source. A key the code
 * merely generates at runtime never sets it.
 */
export function privateKeyInSource(asset: CryptoAsset): boolean {
  return asset.attributes?.private_key_in_source === true;
}

/** assetRowKey is a stable React key: the asset key when the backend sends one. */
export function assetRowKey(asset: CryptoAsset, index: number): string {
  return asset.asset_key || asset.component_key || `${asset.asset_type}:${asset.name}:${index}`;
}

export interface PqcSummary {
  /** What the cell shows: the migration target, e.g. `ML-KEM (FIPS 203, lattice)`. */
  label: string;
  /** The whole recommendation, for the tooltip. */
  full: string;
}

/**
 * pqcSummary splits a recommendation into the migration target a reader scans
 * the column for and the explanation behind it. The analysis writes
 * `<target>: <explanation>`; rendered whole, the explanation made every
 * quantum-vulnerable row several times taller than the rest (live, 2026-09-11).
 * Nothing is dropped — the full text stays one hover away.
 */
export function pqcSummary(text: string | null | undefined): PqcSummary | null {
  const full = (text ?? '').trim();
  if (!full) return null;
  const cut = full.indexOf(': ');
  return { label: cut > 0 ? full.slice(0, cut) : full, full };
}
